composites/Plugin/SnippetEditor/replaceText.js

```diff
@@ -0,0 +1,115 @@
+};
+
+/**
+ * Gets the block of the start of the current Draft.js selection.
+ *
+ * This is a tiny helper function to get the Draft.js selection anchor key and
+ * then the block of that key.
+ * Read more about the Draft.js selection state: https://draftjs.org/docs/api-reference-selection-state.html
+ * Read more about the Draft.js content block: https://draftjs.org/docs/api-reference-content-block.html
+ *
+ * @param {ContentState}   contentState   The Draft.js content state.
+ * @param {SelectionState} selectionState The Draft.js selection state.
+ *
+ * @returns {ContentBlock} The Draft.js ContentBlock.
+ */
+export const getAnchorBlock = ( contentState, selectionState ) => {
+	const key = selectionState.getAnchorKey();
+	return contentState.getBlockForKey( key );
+};
+
+/**
+ * Inserts text into the editor state.
+ *
+ * @param {EditorState} editorState The Draft.js editor state.
+ * @param {string}      text        The text to insert.
+ *
+ * @returns {EditorState} The new editor state.
+ */
+export const insertText = ( editorState, text ) => {
+	const contentState = editorState.getCurrentContent();
+	const selectionState = editorState.getSelection();
+
+	// If there is a selection instead of a caret do nothing.
+	if ( ! selectionState.isCollapsed() ) {
+		return editorState;
+	}
+
+	const newContentState = Modifier.insertText( contentState, selectionState, text );
+	return EditorState.push( editorState, newContentState, "insert-characters" );
+};
+
+/**
+ * Removes the selected text in the editor state.
+ *
+ * @param {EditorState} editorState The Draft.js editor state.
+ *
+ * @returns {EditorState} The new editor state.
+ */
+export const removeSelectedText = ( editorState ) => {
+	const contentState = editorState.getCurrentContent();
+	const selectionState = editorState.getSelection();
+
+	const newContent = Modifier.removeRange( contentState, selectionState, "backward" );
+	return EditorState.push( editorState, newContent, "remove-range" );
+};
+
+/**
+ * Moves the caret to the index.
+ *
+ * @param {EditorState} editorState The Draft.js editor state,
+ * @param {number}      caretIndex  The index of the caret to be.
+ * @param {string}      blockKey    The key for the block to move the caret too. Defaults to current.
+ *
+ * @returns {EditorState} The new editor state.
+ */
+export const moveCaret = ( editorState, caretIndex, blockKey = "" ) => {
+	const contentState = editorState.getCurrentContent();
+	const selectionState = editorState.getSelection();
+
+	// Default to the block where the anchor is currently at.
+	if ( blockKey === "" ) {
+		blockKey = getAnchorBlock( contentState, selectionState ).getKey();
+	}
+
+	const newSelectionState = SelectionState.createEmpty( blockKey )
+	                                   .merge( {
+		                                   anchorOffset: caretIndex,
+		                                   focusOffset: caretIndex,
+	                                   } );
```

Comment:
Ideally, we should avoid this kind of indentation made of spaces or mixes of tabs and spaces.